Question: AUv3 Support

Dear Pedalboard Team, is there a possibility to load AUv3 Plug-Ins with Pedalboard ?

I am getting this error message:

```shell
Traceback (most recent call last):
  File "/Volumes/K2TB/Work/Coding/Python/PedalboardHost/src/host.py", line 9, in <module>
    plugin = load_plugin('/Users/dirk.schiller/Downloads/Builds/auv3.appex')
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/K2TB/Work/Coding/Python/PedalboardHost/.venv/lib/python3.12/site-packages/pedalboard/_pedalboard.py", line 796, in load_plugin
    raise ImportError(
ImportError: Failed to load plugin as VST3Plugin or AudioUnitPlugin. Errors were:
	VST3Plugin: Unable to load plugin /Users/dirk.schiller/Downloads/Builds/auv3.appex: unsupported plugin format or load failure.
	AudioUnitPlugin: Unable to load plugin /Users/dirk.schiller/Downloads/Builds/auv3.appex: unsupported plugin format or load failure. macOS requires plugin files to be moved to /Library/Audio/Plug-Ins/Components/ or ~/Library/Audio/Plug-Ins/Components/ before loading.
```

Also tried to copy the AUv3 Plug-In to the suggested folder, but same issue:

```shell
Traceback (most recent call last):
  File "/Volumes/K2TB/Work/Coding/Python/PedalboardHost/src/host.py", line 9, in <module>
    plugin = load_plugin('/Library/Audio/Plug-Ins/Components/auv3.appex')
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/K2TB/Work/Coding/Python/PedalboardHost/.venv/lib/python3.12/site-packages/pedalboard/_pedalboard.py", line 796, in load_plugin
    raise ImportError(
ImportError: Failed to load plugin as VST3Plugin or AudioUnitPlugin. Errors were:
	VST3Plugin: Unable to load plugin /Library/Audio/Plug-Ins/Components/auv3.appex: unsupported plugin format or load failure.
	AudioUnitPlugin: Unable to load plugin /Library/Audio/Plug-Ins/Components/auv3.appex: unsupported plugin format or load failure.
```

Maybe AUv3 is not supported yet, but cannot find any explicit information in your documentation.

I created a feature branch and try to build pedalboard locally

```shell
git clone https://github.com/spotify/pedalboard.git .
python3 -m venv .venv
source .venv/bin/activate
python -m pip install --upgrade pip
pip install -r test-requirements.txt
```

I needed to download and install JUCE via

```shell
curl -LJO https://github.com/juce-framework/JUCE/releases/download/7.0.11/juce-7.0.11-osx.zip
tar -xf juce-7.0.11-osx.zip -o JUCE
rm juce-7.0.11-osx.zip
```

Other dependencies like Flac also missing:

```shell
python setup.py develop

...
In file included from /Volumes/K2TB/Work/Coding/Python/Pedalboard/feature.add-auv3/pedalboard/juce_overrides/juce_PatchedFLACAudioFormat.cpp:139:
/Volumes/K2TB/Work/Coding/Python/Pedalboard/feature.add-auv3/pedalboard/juce_overrides/../../JUCE/modules/juce_audio_formats/codecs/flac/libFLAC/format.c:57:45: error: use of undeclared identifier 'PACKAGE_VERSION'
FLAC_API const char *FLAC__VERSION_STRING = PACKAGE_VERSION;
                                            ^
/Volumes/K2TB/Work/Coding/Python/Pedalboard/feature.add-auv3/pedalboard/juce_overrides/../../JUCE/modules/juce_audio_formats/codecs/flac/libFLAC/format.c:58:64: error: expected ';' after top level declarator
FLAC_API const char *FLAC__VENDOR_STRING = "reference libFLAC " PACKAGE_VERSION " 20230623";
                                                               ^
                                                               ;
In file included from /Volumes/K2TB/Work/Coding/Python/Pedalboard/feature.add-auv3/pedalboard/juce_overrides/juce_PatchedFLACAudioFormat.cpp:147:
/Volumes/K2TB/Work/Coding/Python/Pedalboard/feature.add-auv3/pedalboard/juce_overrides/../../vendors/libFLAC/metadata_object.c:41:10: fatal error: '../../JUCE/modules/juce_audio_formats/codecs/flac/libFLAC/include/private/metadata.h' file not found
#include "../../JUCE/modules/juce_audio_formats/codecs/flac/libFLAC/include/private/metadata.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3 errors generated.
error: command '/usr/bin/clang' failed with exit code 1
```

How do you install the necessary dependencies ?

--- a/pedalboard/AudioUnitParser.mm
+++ b/pedalboard/AudioUnitParser.mm
@@ -19,6 +19,7 @@
 
 #if JUCE_MAC
 #include <AudioToolbox/AudioUnitUtilities.h>
+#include <AVFoundation/AVFAudio.h>
 
 namespace Pedalboard {
 
@@ -82,24 +83,38 @@ getAudioUnitIdentifiersFromFile(const juce::String &filename) {
     return identifiers;
   }
 
-  // Use NSBundle to open and extract identifiers:
-  NSBundle *bundle =
-      [[NSBundle alloc] initWithPath:(NSString *)filename.toCFString()];
+  if (file.hasFileExtension(".appex")) {
+    AVAudioUnitComponentManager *manager = [AVAudioUnitComponentManager sharedAudioUnitComponentManager];
+    AudioComponentDescription desc = {0};
+    NSArray *audioComponents = [manager componentsMatchingDescription:desc];
+    NSString *filePathWithoutTrailingSlash = (NSString *)(filename.trimCharactersAtEnd("/") + "/").toCFString();
 
-  NSArray *audioComponents =
-      [bundle objectForInfoDictionaryKey:@"AudioComponents"];
-  for (NSDictionary *component in audioComponents) {
-    AudioComponentDescription desc;
-    desc.componentManufacturer = stringToOSType(
-        nsStringToJuce((NSString *)[component valueForKey:@"manufacturer"]));
-    desc.componentType = stringToOSType(
-        nsStringToJuce((NSString *)[component valueForKey:@"type"]));
-    desc.componentSubType = stringToOSType(
-        nsStringToJuce((NSString *)[component valueForKey:@"subtype"]));
-    identifiers.push_back(createPluginIdentifier(desc).toStdString());
+    for (AVAudioUnitComponent *component in audioComponents) {
+      if ([component.componentURL.absoluteString hasSuffix:filePathWithoutTrailingSlash]) {
+        identifiers.push_back(createPluginIdentifier(component.audioComponentDescription).toStdString());
+        break;
+      }
+    }
+  } else {
+    // Use NSBundle to open and extract identifiers:
+    NSBundle *bundle =
+        [[NSBundle alloc] initWithPath:(NSString *)filename.toCFString()];
+
+    NSArray *audioComponents =
+        [bundle objectForInfoDictionaryKey:@"AudioComponents"];
+    for (NSDictionary *component in audioComponents) {
+      AudioComponentDescription desc;
+      desc.componentManufacturer = stringToOSType(
+          nsStringToJuce((NSString *)[component valueForKey:@"manufacturer"]));
+      desc.componentType = stringToOSType(
+          nsStringToJuce((NSString *)[component valueForKey:@"type"]));
+      desc.componentSubType = stringToOSType(
+          nsStringToJuce((NSString *)[component valueForKey:@"subtype"]));
+      identifiers.push_back(createPluginIdentifier(desc).toStdString());
+    }
+
+    [bundle release];
   }
-
-  [bundle release];
 
   return identifiers;
 }

--- a/pedalboard/ExternalPlugin.h
+++ b/pedalboard/ExternalPlugin.h
@@ -301,16 +301,18 @@ private:
 #endif
 
 static bool audioUnitIsInstalled(const juce::String audioUnitFilePath) {
-  return audioUnitFilePath.contains("/Library/Audio/Plug-Ins/Components/");
+  return !audioUnitFilePath.endsWith(".appex") &&
+         !audioUnitFilePath.endsWith(".appex/") &&
+         audioUnitFilePath.contains("/Library/Audio/Plug-Ins/Components/");
 }
 
 template <typename ExternalPluginType>
-static std::vector<std::string> getPluginNamesForFile(std::string filename) {
+static juce::OwnedArray<juce::PluginDescription>
+scanPluginDescriptions(std::string filename) {
   juce::MessageManager::getInstance();
-
   ExternalPluginType format;
-  juce::OwnedArray<juce::PluginDescription> typesFound;
 
+  juce::OwnedArray<juce::PluginDescription> typesFound;
   std::string errorMessage = "Unable to scan plugin " + filename +
                              ": unsupported plugin format or scan failure.";
 
@@ -320,7 +322,34 @@ static std::vector<std::string> getPluginNamesForFile(std::string filename) {
     auto identifiers = getAudioUnitIdentifiersFromFile(filename);
     // For each plugin in the identified bundle, scan using its AU identifier:
     for (int i = 0; i < identifiers.size(); i++) {
-      format.findAllTypesForFile(typesFound, identifiers[i]);
+      std::string identifier = identifiers[i];
+
+      juce::String pluginName, version, manufacturer;
+      AudioComponentDescription componentDesc;
+
+      bool needsAsyncInstantiation =
+          juce::String(filename).endsWith(".appex") ||
+          juce::String(filename).endsWith(".appex/");
+      if (needsAsyncInstantiation &&
+          juce::MessageManager::getInstance()->isThisTheMessageThread()) {
+        // We can't scan AUv3 plugins synchronously, so we have to pump the
+        // message thread and wait for the scan to complete on another thread.
+        bool done = false;
+        std::thread thread =
+            std::thread([&identifier, &typesFound, &format, &done]() {
+              format.findAllTypesForFile(typesFound, identifier);
+              done = true;
+            });
+
+        // Pump the message thread until the scan is complete:
+        while (!done) {
+          juce::MessageManager::getInstance()->runDispatchLoopUntil(1);
+        }
+
+        thread.join();
+      } else {
+        format.findAllTypesForFile(typesFound, identifiers[i]);
+      }
     }
 
     if (typesFound.isEmpty() && !audioUnitIsInstalled(filename)) {
@@ -336,6 +365,14 @@ static std::vector<std::string> getPluginNamesForFile(std::string filename) {
   if (typesFound.isEmpty()) {
     throw pybind11::import_error(errorMessage);
   }
+
+  return typesFound;
+}
+
+template <typename ExternalPluginType>
+static std::vector<std::string> getPluginNamesForFile(std::string filename) {
+  juce::OwnedArray<juce::PluginDescription> typesFound =
+      scanPluginDescriptions<ExternalPluginType>(filename);
 
   std::vector<std::string> pluginNames;
   for (int i = 0; i < typesFound.size(); i++) {
@@ -492,9 +529,6 @@ public:
     // Without this, we get an assert(false) from JUCE at runtime
     juce::MessageManager::getInstance();
 
-    juce::OwnedArray<juce::PluginDescription> typesFound;
-    ExternalPluginType format;
-
     pluginFormatManager.addDefaultFormats();
     pluginFormatManager.addFormat(new juce::PatchedVST3PluginFormat());
 
@@ -509,20 +543,9 @@ public:
                                    ": plugin file not found.");
     }
 
-#if JUCE_PLUGINHOST_AU && JUCE_MAC
-    if constexpr (std::is_same<ExternalPluginType,
-                               juce::AudioUnitPluginFormat>::value) {
-      auto identifiers = getAudioUnitIdentifiersFromFile(pluginFileStripped);
-      // For each plugin in the identified bundle, scan using its AU identifier:
-      for (int i = 0; i < identifiers.size(); i++) {
-        format.findAllTypesForFile(typesFound, identifiers[i]);
-      }
-    } else {
-      format.findAllTypesForFile(typesFound, pluginFileStripped);
-    }
-#else
-    format.findAllTypesForFile(typesFound, pluginFileStripped);
-#endif
+    juce::OwnedArray<juce::PluginDescription> typesFound =
+        scanPluginDescriptions<ExternalPluginType>(
+            pluginFileStripped.toStdString());
 
     if (!typesFound.isEmpty()) {
       if (typesFound.size() == 1) {
@@ -681,9 +704,9 @@ public:
     {
       std::lock_guard<std::mutex> lock(EXTERNAL_PLUGIN_MUTEX);
 
-      pluginInstance = pluginFormatManager.createPluginInstance(
-          foundPluginDescription, ExternalLoadSampleRate,
-          ExternalLoadMaximumBlockSize, loadError);
+      pluginInstance =
+          createPluginInstance(foundPluginDescription, ExternalLoadSampleRate,
+                               ExternalLoadMaximumBlockSize, loadError);
 
       if (!pluginInstance) {
         throw pybind11::import_error("Unable to load plugin " +
@@ -709,7 +732,7 @@ public:
         if (reloadType == ExternalPluginReloadType::PersistsAudioOnReset) {
           // Reload again, as we just passed audio into a plugin that
           // we know doesn't reset itself cleanly!
-          pluginInstance = pluginFormatManager.createPluginInstance(
+          pluginInstance = createPluginInstance(
               foundPluginDescription, ExternalLoadSampleRate,
               ExternalLoadMaximumBlockSize, loadError);
 
@@ -1315,6 +1338,34 @@ public:
   juce::PluginDescription foundPluginDescription;
 
 private:
+  std::unique_ptr<juce::AudioPluginInstance>
+  createPluginInstance(const juce::PluginDescription &foundPluginDescription,
+                       double rate, int blockSize, juce::String &loadError) {
+    std::unique_ptr<juce::AudioPluginInstance> instance;
+    instance = pluginFormatManager.createPluginInstance(
+        foundPluginDescription, rate, blockSize, loadError);
+    if (!instance && loadError.contains(
+                         "This plug-in cannot be instantiated synchronously")) {
+      bool done = false;
+      std::thread thread =
+          std::thread([this, &instance, &foundPluginDescription, &rate,
+                       &blockSize, &loadError, &done]() {
+            instance = pluginFormatManager.createPluginInstance(
+                foundPluginDescription, rate, blockSize, loadError);
+            done = true;
+          });
+
+      // Pump the message thread until the scan is complete:
+      while (!done) {
+        juce::MessageManager::getInstance()->runDispatchLoopUntil(1);
+      }
+
+      thread.join();
+    }
+
+    return instance;
+  }
+
   constexpr static int ExternalLoadSampleRate = 44100,
                        ExternalLoadMaximumBlockSize = 8192;
   juce::String pathToPluginFile;

